diff --git a/torch/_dynamo/compiled_autograd.py b/torch/_dynamo/compiled_autograd.py
index 582c20b5995..5955c4e4170 100644
--- a/torch/_dynamo/compiled_autograd.py
+++ b/torch/_dynamo/compiled_autograd.py
@@ -57,6 +57,7 @@ from torch.fx.experimental.symbolic_shapes import DimDynamic, ShapeEnv
 from torch.fx.traceback import preserve_node_meta, set_stack_trace
 from torch.utils._ordered_set import OrderedSet
 from torch.utils._traceback import CapturedTraceback
+from torch.distributed._functional_collectives import AsyncCollectiveTensor


 if TYPE_CHECKING:
@@ -66,6 +67,8 @@ if TYPE_CHECKING:
 compiled_autograd_log = getArtifactLogger(__name__, "compiled_autograd")
 verbose_log = getArtifactLogger(__name__, "compiled_autograd_verbose")

+param_to_optim = None
+

 def snapshot_verbose_logging_enabled():
     return torch._logging._internal.log_state.is_artifact_enabled(
@@ -281,6 +284,19 @@ class AutogradCompilerInstance:
         self.aot_id_counter: dict[int, int] = defaultdict(int)
         self.compile_context = make_compile_context(self.id)
         self.compile_context.__enter__()
+        if param_to_optim:
+            logs = {}
+            for i,inp in enumerate(inputs):
+                if id(inp) in param_to_optim:
+                    logs[i] = f"{param_to_optim[id(inp)]}"
+            trace_structured(
+                "artifact",
+                metadata_fn=lambda: {
+                    "name": "param_to_optim_ca",
+                    "encoding": "json",
+                },
+                payload_fn=lambda: logs,
+            )
         self.nan_checker = NaNChecker(accumulate_grad) if check_nans else None
         self.start_time_ns = time.time_ns()
         get_chromium_event_logger().log_event_start(
@@ -933,6 +949,17 @@ class AutogradCompilerInstance:
         if self.nan_checker:
             self.nan_checker.prep_with_graph(self.fx_tracer.graph)

+        param_runtime_idx = []
+        for node in self.fx_tracer.graph.find_nodes(
+            op="call_function", target=torch.ops.inductor.accumulate_grad_.default
+        ):
+            assert "getitem" in node.args[0].name
+            assert len(node.args[0].args) == 2
+            assert node.args[0].args[0].name == "inputs"
+            idx = node.args[0].args[1]
+            assert isinstance(idx, int)
+            param_runtime_idx.append(idx)
+
         graph = self.create_graph_module(f"CompiledAutograd{self.id}")
         set_locals_to_steal(graph, ["inputs"])
         lazy_graph_code = lazy_format_graph_code(
@@ -951,6 +978,9 @@ class AutogradCompilerInstance:

         def runtime_wrapper(compiled_fn, inputs, sizes, scalars, hooks, packed_inputs):
             global in_compiled_autograd_region
+            params = []
+            for i in param_runtime_idx:
+                params.append(inputs[i])
             try:
                 in_compiled_autograd_region = True
                 if self.nan_checker:
@@ -966,6 +996,8 @@ class AutogradCompilerInstance:
                     return out
             finally:
                 in_compiled_autograd_region = False
+                for p in params:
+                    p.grad = AsyncCollectiveTensor(p.grad)

         get_chromium_event_logger().log_event_end(
             "compiled_autograd",
diff --git a/torch/_inductor/codecache.py b/torch/_inductor/codecache.py
index 2ee7f3fdc2f..18ddd4719ea 100644
--- a/torch/_inductor/codecache.py
+++ b/torch/_inductor/codecache.py
@@ -1088,7 +1088,7 @@ class FxGraphCache:
         metrics.CachedMetricsHelper.apply_deltas(graph.metrics_deltas)
         counters["inductor"] += graph.counter_deltas

-        output_code_log.debug("Output code: \n%s", code)
+        # output_code_log.debug("Output code: \n%s", code)
         output_code_log.debug("Output code written to: %s", artifact_path)
         # On cache hit, use artifact path as filename
         trace_structured(
diff --git a/torch/_inductor/config.py b/torch/_inductor/config.py
index 9bdc5b7232c..58028bec056 100644
--- a/torch/_inductor/config.py
+++ b/torch/_inductor/config.py
@@ -643,7 +643,9 @@ worker_start_method: str = decide_worker_start_method()
 # Flags to turn on all_reduce fusion. These 2 flags should be automaticaly turned
 # on by DDP and should not be set by the users.
 _fuse_ddp_communication = False
-_fuse_ddp_bucket_size = 25
+# import sys
+# _fuse_ddp_bucket_size = sys.maxsize
+_fuse_ddp_bucket_size = 0

 # Flag to control which fusion passes to apply. Functions in the list will
 # be applied in order. There are two different different fusion passes
@@ -660,7 +662,7 @@ _fuse_ddp_bucket_size = 25
 # "schedule_comm_wait" in the future and remove the one here.
 _fuse_ddp_communication_passes: list[Union[Callable[..., None], str]] = [
     "fuse_ddp_with_concat_op",
-    "schedule_comm_wait",
+    # "schedule_comm_wait",
 ]

 _micro_pipeline_tp: bool = False
diff --git a/torch/_inductor/fx_passes/ddp_fusion.py b/torch/_inductor/fx_passes/ddp_fusion.py
index 2d9409523c1..445e05be56e 100644
--- a/torch/_inductor/fx_passes/ddp_fusion.py
+++ b/torch/_inductor/fx_passes/ddp_fusion.py
@@ -475,19 +475,20 @@ def _fuse_ddp_communication(
         ):
             return False

-        if len(block.wait_nodes[0].users) != 1:
-            # gradient/wait node should only be used by one user
-            return False
-
-        # Two cases:
-        # 1. gradient/wait node should be directly used by the output
-        # if gradient is None before bwd.
-        # 2. gradient/wait node should be directly used by copy_.
-        if (
-            output not in block.wait_nodes[0].users
-            and next(iter(block.wait_nodes[0].users)).target != aten.copy_.default
-        ):
-            return False
+        # wtf no users
+        # if len(block.wait_nodes[0].users) != 1:
+        #     # gradient/wait node should only be used by one user
+        #     return False
+
+        # # Two cases:
+        # # 1. gradient/wait node should be directly used by the output
+        # # if gradient is None before bwd.
+        # # 2. gradient/wait node should be directly used by copy_.
+        # if (
+        #     output not in block.wait_nodes[0].users
+        #     and next(iter(block.wait_nodes[0].users)).target != aten.copy_.default
+        # ):
+        #     return False

         return True

@@ -498,6 +499,8 @@ def _fuse_ddp_communication(
     comm_blocks = get_all_comm_blocks(graph, ops, comm_filter=ddp_reducer_filter)
     node_indices = {node: i for i, node in enumerate(graph.nodes)}

+    # if torch._dynamo.compiled_autograd.in_compiled_autograd_region:
+    #     breakpoint()
     for block in algorithm_fn(graph, comm_blocks):
         fusion_fn(graph, block, node_indices)

diff --git a/torch/_inductor/fx_passes/post_grad.py b/torch/_inductor/fx_passes/post_grad.py
index 17fb95fba3c..e9c3c495c54 100644
--- a/torch/_inductor/fx_passes/post_grad.py
+++ b/torch/_inductor/fx_passes/post_grad.py
@@ -92,9 +92,29 @@ def post_grad_passes(gm: torch.fx.GraphModule, is_inference: bool):
         gm.graph.eliminate_dead_code()

     if is_inference and config.reorder_for_locality:
+        trace_structured(
+            "artifact",
+            metadata_fn=lambda: {
+                "name": "before_reorder_for_locality",
+                "encoding": "string",
+            },
+            payload_fn=lambda: gm.print_readable(
+                print_output=False, include_stride=True, include_device=True
+            ),
+        )
         GraphTransformObserver(gm, "reorder_for_locality").apply_graph_pass(
             reorder_for_locality
         )
+        trace_structured(
+            "artifact",
+            metadata_fn=lambda: {
+                "name": "after_reorder_for_locality",
+                "encoding": "string",
+            },
+            payload_fn=lambda: gm.print_readable(
+                print_output=False, include_stride=True, include_device=True
+            ),
+        )

     fake_tensor_updater = FakeTensorUpdater(gm.graph)

@@ -491,6 +511,8 @@ def reorder_for_locality(graph: torch.fx.Graph):
             and all((n in seen_nodes) for n in other_node.users)
             and get_mutation_region_id(graph, node)
             == get_mutation_region_id(graph, other_node)
+            # and node.meta["collective_region_id"]
+            # == other_node.meta["collective_region_id"]
         ):
             if node.target == torch.ops._c10d_functional.wait_tensor.default:
                 # This is a wait node, it's producer is some collective node. When
@@ -505,6 +527,14 @@ def reorder_for_locality(graph: torch.fx.Graph):

     seen_nodes = OrderedSet[torch.fx.Node]()

+    # if torch._dynamo.compiled_autograd.in_compiled_autograd_region:
+    # collective_region_id = 0
+    # for node in graph.nodes:
+    #     if node.op == "call_function" and node.target == torch.ops._c10d_functional.all_reduce.default:
+    #         collective_region_id += 1
+
+    #     node.meta["collective_region_id"] = collective_region_id
+
     # only reorder nodes before the first copy_ in the graph.
     # copy_ will appear at the end of functionalized graphs when there is mutation on inputs,
     # and this reordering doesnt work well with mutation
diff --git a/torch/_inductor/graph.py b/torch/_inductor/graph.py
index bc9ab904494..f9958d637c3 100644
--- a/torch/_inductor/graph.py
+++ b/torch/_inductor/graph.py
@@ -2105,7 +2105,7 @@ class GraphLowering(torch.fx.Interpreter):
             wrapper_code.value = tuning_code + wrapper_code.value
         if GraphLowering.save_output_code is not None:
             GraphLowering.save_output_code(wrapper_code.value)
-        output_code_log.debug("Output code: \n%s", wrapper_code.value)
+        # output_code_log.debug("Output code: \n%s", wrapper_code.value)

         inductor_meta = autotune_cache.inductor_meta_from_config()
         AutotuneCacheBundler.begin_compile(inductor_meta, code=wrapper_code.value)
diff --git a/torch/fx/proxy.py b/torch/fx/proxy.py
index ce1814dd7f2..2dba70e8131 100644
--- a/torch/fx/proxy.py
+++ b/torch/fx/proxy.py
@@ -247,6 +247,9 @@ class TracerBase:

         if self.record_stack_traces and not proxy.node.stack_trace:
             proxy.node.stack_trace = "".join(CapturedTraceback.extract().format())
+        elif torch.fx._symbolic_trace.is_fx_tracing() and not proxy.node.stack_trace:
+            # under symbolic_trace
+            proxy.node.stack_trace = "".join(CapturedTraceback.extract().format()[:-2])

         return proxy

diff --git a/torch/optim/adam.py b/torch/optim/adam.py
index 2f01b1d683b..955366d1e96 100644
--- a/torch/optim/adam.py
+++ b/torch/optim/adam.py
@@ -155,7 +155,6 @@ class Adam(Optimizer):
                         "Adam does not support sparse gradients, please consider SparseAdam instead"
                     )
                 grads.append(p.grad)
-
                 state = self.state[p]
                 # Lazy state initialization
                 if len(state) == 0:
